### PR TITLE
Backport of sam/abci-responses (#9090)

### PR DESCRIPTION
**Pull request by samricotta**
Backport of `sam/abci-responses` https://github.com/tendermint/tendermint/pull/9090

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -405,6 +405,11 @@ type RPCConfig struct {
 
 	// pprof listen address (https://golang.org/pkg/net/http/pprof)
 	PprofListenAddress string `mapstructure:"pprof_laddr"`
+
+	// Set false to ensure ABCI responses are persisted.
+	// ABCI responses are required for /BlockResults RPC queries, and
+	// to reindex events in the command-line tool.
+	DiscardABCIResponses bool `mapstructure:"persist-abci-responses"`
```
Comment (williambanfield) [inline config/config.go:412]:
```suggestion
	// Set false to ensure ABCI responses are persisted.
	// ABCI responses are required for /BlockResults RPC queries, and
	// to reindex events in the command-line tool.
	DiscardABCIResponses bool `mapstructure:"discard_abci_responses"`
```

The keys changed from `this_style` to `this-style` between releases. I'm also realizing that the config key is `persist-abci-responses` but the field name is `DiscardABCIResponses` which have opposite meanings. I think we should change the key to match the field name.

**Review by williambanfield — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 1 above)
Comment (samricotta) [reply config/config.go:412]: Yeah this completely makes sense, Ill make those updates

**Review by williambanfield — APPROVED** [review]
Looks like it needs to have `gofmt` run in a few spots. Should be good after that.

File: state/state.go [hunk 2]
```diff
@@ -17,10 +17,7 @@ import (
 	"github.com/tendermint/tendermint/version"
 )
 
-// database keys
-var (
-	stateKey = []byte("stateKey")
-)
+// database key
```
Comment (cmwaters) [inline state/state.go:20]: what happened here?

File: state/store.go [hunk 3]
```diff
@@ -23,18 +24,49 @@ const (
 	valSetCheckpointInterval = 100000
 )
 
+const (
+	prefixValidators       = int64(5)
+	prefixConsensusParams  = int64(6)
+	prefixAllABCIResponses = int64(7)
+	prefixState            = int64(8)
+	prefixLastABCIResponse = int64(13)
+)
```
Comment (cmwaters) [inline state/store.go:33]: You need to use all the old prefixes and key encodings

Comment (cmwaters) [reply state/store.go:33]: Else this breaks v0.34.x

File: state/store.go [hunk 4]
```diff
@@ -385,12 +431,43 @@ func (store dbStore) LoadABCIResponses(height int64) (*tmstate.ABCIResponses, er
...
+// This method is used for recovering in the case that we called the Commit ABCI
+// method on the application but crashed before persisting the results.
+func (store dbStore) LoadLastABCIResponse(height int64) (*tmstate.ABCIResponses, error) {
+	bz, err := store.db.Get(lastABCIResponseKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bz) == 0 {
+		return nil, errors.New("value retrieved from db is empty")
```
Comment (cmwaters) [inline state/store.go:447]:
```suggestion
		return nil, errors.New("no last ABCI response has been persisted")
```

File: state/store.go (hunk 3 above)
Comment (cmwaters) [reply state/store.go:33]: We should probably even have tests that ensure keys don't change throughout a release

**Review by cmwaters — CHANGES_REQUESTED** [review]

**Comment by williambanfield** [comment]
Should we instead be porting this into `main` so that it goes out with the v0.37 release, or do we need it in v0.34 quickly as well?

File: state/store.go [hunk 5]
```diff
@@ -75,13 +81,24 @@ type Store interface {
...
 type dbStore struct {
 	db dbm.DB
+
+	StoreOptions
+}
+
+type StoreOptions struct {
+
+	// DiscardABCIResponses determines whether or not the store
+	// retains all ABCIResponses. If DiscardABCiResponses is enabled,
```
Comment (cmwaters) [inline state/store.go:91]:
```suggestion
	// retains all ABCIResponses. If DiscardABCIResponses is enabled,
```

File: state/store.go [hunk 6]
```diff
@@ -385,12 +403,43 @@ func (store dbStore) LoadABCIResponses(height int64) (*tmstate.ABCIResponses, er
...
+	return abciResponse.AbciResponses, nil
+}
+
 // SaveABCIResponses persists the ABCIResponses to the database.
 // This is useful in case we crash after app.Commit and before s.Save().
 // Responses are indexed by height so they can also be loaded later to produce
 // Merkle proofs.
 //
-// Exposed for testing.
+//CONTRACT: height must be monotonically increasing every time this is called.
```
Comment (cmwaters) [inline state/store.go:442]:
```suggestion
// CONTRACT: height must be monotonically increasing every time this is called.
```

**Review by cmwaters — APPROVED** [review]
Looks good. Just a few nits

**Pull request merged**